File: libgomp/testsuite/libgomp.c-c++-common/target-is-accessible-2.c
/* { dg-do run } */
/* { dg-require-effective-target omp_usm } */

#include <omp.h>

#pragma omp requires unified_shared_memory

int
main ()
{
  int *a = (int *) omp_alloc (sizeof (int), ompx_unified_shared_mem_alloc);
  if (!a)
    __builtin_abort ();

  for (int d = 0; d <= omp_get_num_devices (); d++)
    if (!omp_target_is_accessible (a, sizeof (int), d))
      __builtin_abort ();

  omp_free(a, ompx_unified_shared_mem_alloc);
  return 0;
}

File: libgomp/testsuite/libgomp.fortran/target-is-accessible-2.f90
! { dg-do run }
! { dg-require-effective-target omp_usm }

program main
  use omp_lib
  use iso_c_binding
  implicit none (external, type)
  integer :: d
  type(c_ptr) :: p

  !$omp requires unified_shared_memory

  p = omp_alloc (sizeof (d), ompx_unified_shared_mem_alloc)
  if (.not. c_associated (p)) stop 1

  do d = 0, omp_get_num_devices ()
    if (omp_target_is_accessible (p, c_sizeof (d), d) == 0) &
      stop 2;
  end do

  call omp_free (p, ompx_unified_shared_mem_alloc);
end program main
